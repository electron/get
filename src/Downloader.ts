export interface Downloader<T> {
  download(url: string, targetFilePath: string, options: T): Promise<void>;
}
