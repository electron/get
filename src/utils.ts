import * as childProcess from 'child_process';
import * as fs from 'fs-extra';
import * as os from 'os';
import * as path from 'path';

async function useAndRemoveDirectory<T>(
  directory: string,
  fn: (directory: string) => Promise<T>,
): Promise<T> {
  let result: T;
  try {
    result = await fn(directory);
  } finally {
    await fs.remove(directory);
  }
  return result;
}

export async function withTempDirectoryIn<T>(
  parentDirectory: string = os.tmpdir(),
  fn: (directory: string) => Promise<T>,
): Promise<T> {
  const tempDirectoryPrefix = 'electron-download-';
  const tempDirectory = await fs.mkdtemp(path.resolve(parentDirectory, tempDirectoryPrefix));
  return useAndRemoveDirectory(tempDirectory, fn);
}

export async function withTempDirectory<T>(fn: (directory: string) => Promise<T>): Promise<T> {
  return withTempDirectoryIn(undefined, fn);
}

export function normalizeVersion(version: string): string {
  if (!version.startsWith('v')) {
    return `v${version}`;
  }
  return version;
}

/**
 * Runs the `uname` command and returns the trimmed output.
 */
export function uname(): string {
  return childProcess
    .execSync('uname -m')
    .toString()
    .trim();
}

/**
 * Generates an architecture name that would be used in an Electron or Node.js
 * download file name.
 */
export function getNodeArch(arch: string): string {
  if (arch === 'arm') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    switch ((process.config.variables as any).arm_version) {
      case '6':
        return uname();
      case '7':
      default:
        return 'armv7l';
    }
  }

  return arch;
}

/**
 * Generates an architecture name that would be used in an Electron or Node.js
 * download file name, from the `process` module information.
 */
export function getHostArch(): string {
  return getNodeArch(process.arch);
}

export function ensureIsTruthyString<T, K extends keyof T>(obj: T, key: K): void {
  if (!obj[key] || typeof obj[key] !== 'string') {
    throw new Error(`Expected property "${key}" to be provided as a string but it was not`);
  }
}

export function isOfficialLinuxIA32Download(
  platform: string,
  arch: string,
  version: string,
  mirrorOptions?: object,
): boolean {
  return (
    platform === 'linux' &&
    arch === 'ia32' &&
    Number(version.slice(1).split('.')[0]) >= 4 &&
    typeof mirrorOptions === 'undefined'
  );
}

/**
 * Find the value of a environment variable which may or may not have the
 * prefix, in a case-insensitive manner.
 */
export function getEnv(prefix = ''): (name: string) => string | undefined {
  const envsLowerCase: NodeJS.ProcessEnv = {};

  for (const envKey in process.env) {
    envsLowerCase[envKey.toLowerCase()] = process.env[envKey];
  }

  return (name: string): string | undefined => {
    return (
      envsLowerCase[`${prefix}${name}`.toLowerCase()] ||
      envsLowerCase[name.toLowerCase()] ||
      undefined
    );
  };
}

export function setEnv(key: string, value: string | undefined): void {
  // The `void` operator always returns `undefined`.
  // See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void
  if (value !== void 0) {
    process.env[key] = value;
  }
}
