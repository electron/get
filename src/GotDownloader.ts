/* eslint-disable */
// @ts-ignore - require(esm) supported in Node 22.12
import got, { HTTPError, Progress as GotProgress, Options as GotOptions } from 'got';
/* eslint-enable */

import fs from 'node:fs';
import path from 'node:path';
import ProgressBar from 'progress';

import { Downloader } from './Downloader';

const PROGRESS_BAR_DELAY_IN_SECONDS = 30;

/**
 * Options for the default [`got`](https://github.com/sindresorhus/got) Downloader implementation.
 *
 * @category Downloader
 * @see {@link https://github.com/sindresorhus/got/tree/v11.8.5?tab=readme-ov-file#options | `got#options`} for possible keys/values.
 */
export type GotDownloaderOptions = GotOptions & { isStream?: true } & {
  /**
   * if defined, triggers every time `got`'s
   * {@link https://github.com/sindresorhus/got/tree/v11.8.5?tab=readme-ov-file#downloadprogress | `downloadProgress``} event callback is triggered.
   */
  getProgressCallback?: (progress: GotProgress) => Promise<void>;
  /**
   * if `true`, disables the console progress bar (setting the `ELECTRON_GET_NO_PROGRESS`
   * environment variable to a non-empty value also does this).
   */
  quiet?: boolean;
};

/**
 * Default {@link Downloader} implemented with {@link https://npmjs.com/package/got | `got`}.
 * @category Downloader
 */
export class GotDownloader implements Downloader<GotDownloaderOptions> {
  async download(
    url: string,
    targetFilePath: string,
    options?: Partial<GotDownloaderOptions>,
  ): Promise<void> {
    if (!options) {
      options = {};
    }
    const { quiet, getProgressCallback, ...gotOptions } = options;
    let downloadCompleted = false;
    let bar: ProgressBar | undefined;
    let progressPercent: number;
    let timeout: NodeJS.Timeout | undefined = undefined;
    await fs.promises.mkdir(path.dirname(targetFilePath), { recursive: true });
    const writeStream = fs.createWriteStream(targetFilePath);

    if (!quiet || !process.env.ELECTRON_GET_NO_PROGRESS) {
      const start = new Date();
      timeout = setTimeout(() => {
        if (!downloadCompleted) {
          bar = new ProgressBar(
            `Downloading ${path.basename(url)}: [:bar] :percent ETA: :eta seconds `,
            {
              curr: progressPercent,
              total: 100,
            },
          );
          // https://github.com/visionmedia/node-progress/issues/159
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (bar as any).start = start;
        }
      }, PROGRESS_BAR_DELAY_IN_SECONDS * 1000);
    }
    await new Promise<void>((resolve, reject) => {
      const downloadStream = got.stream(url, gotOptions);
      downloadStream.on('downloadProgress', async (progress) => {
        progressPercent = progress.percent;
        if (bar) {
          bar.update(progress.percent);
        }
        if (getProgressCallback) {
          await getProgressCallback(progress);
        }
      });
      downloadStream.on('error', (error) => {
        if (error instanceof HTTPError && error.response.statusCode === 404) {
          error.message += ` for ${error.response.url}`;
        }
        if (writeStream.destroy) {
          writeStream.destroy(error);
        }

        reject(error);
      });
      writeStream.on('error', (error) => reject(error));
      writeStream.on('close', () => resolve());

      downloadStream.pipe(writeStream);
    });

    downloadCompleted = true;
    if (timeout) {
      clearTimeout(timeout);
    }
  }
}
