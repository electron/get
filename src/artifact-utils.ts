import { ElectronArtifactDetails, MirrorOptions } from './types';
import { ensureIsTruthyString } from './utils';

const BASE_URL = 'https://github.com/electron/electron/releases/download/';
const NIGHTLY_BASE_URL = 'https://github.com/electron/nightlies/releases/download/';

export function getArtifactFileName(details: ElectronArtifactDetails) {
  ensureIsTruthyString(details, 'artifactName');

  if (details.isGeneric) {
    return details.artifactName;
  }

  ensureIsTruthyString(details, 'arch');
  ensureIsTruthyString(details, 'platform');
  ensureIsTruthyString(details, 'version');

  return `${[
    details.artifactName,
    details.version,
    details.platform,
    details.arch,
    ...(details.artifactSuffix ? [details.artifactSuffix] : []),
  ].join('-')}.zip`;
}

function mirrorVar(
  name: keyof Omit<MirrorOptions, 'resolveAssetURL'>,
  options: MirrorOptions,
  defaultValue: string,
) {
  // Convert camelCase to camel_case for env var reading
  const lowerName = name.replace(/([a-z])([A-Z])/g, (_, a, b) => `${a}_${b}`).toLowerCase();

  return (
    process.env[`NPM_CONFIG_ELECTRON_${lowerName.toUpperCase()}`] ||
    process.env[`npm_config_electron_${lowerName}`] ||
    process.env[`npm_package_config_electron_${lowerName}`] ||
    process.env[`ELECTRON_${lowerName.toUpperCase()}`] ||
    options[name] ||
    defaultValue
  );
}

export async function getArtifactRemoteURL(details: ElectronArtifactDetails): Promise<string> {
  const opts: MirrorOptions = details.mirrorOptions || {};
  let base = mirrorVar('mirror', opts, BASE_URL);
  if (details.version.includes('nightly')) {
    base = mirrorVar('nightly_mirror', opts, NIGHTLY_BASE_URL);
  }
  const path = mirrorVar('customDir', opts, details.version).replace(
    '{{ version }}',
    details.version.replace(/^v/, ''),
  );
  const file = mirrorVar('customFilename', opts, getArtifactFileName(details));

  // Allow customized download URL resolution.
  if (opts.resolveAssetURL) {
    const url = await opts.resolveAssetURL(details);
    return url;
  }

  return `${base}${path}/${file}`;
}
