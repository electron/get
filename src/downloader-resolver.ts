import { DownloadOptions } from './types.js';
import { Downloader } from './Downloader.js';

// TODO: Resolve the downloader or default to GotDownloader
// Current thoughts are a dot-file traversal for something like
// ".electron.downloader" which would be a text file with the name of the
// npm module to import() and use as the downloader
import { GotDownloader } from './GotDownloader.js';

export async function getDownloaderForSystem(): Promise<Downloader<DownloadOptions>> {
  return new GotDownloader();
}
