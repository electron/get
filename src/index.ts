import debug from 'debug';
import * as path from 'path';
import * as sumchecker from 'sumchecker';

import { getArtifactFileName, getArtifactRemoteURL } from './artifact-utils';
import {
  ElectronArtifactDetails,
  ElectronDownloadRequestOptions,
  ElectronPlatformArtifactDetailsWithDefaults,
} from './types';
import { Cache } from './Cache';
import { getDownloaderForSystem } from './downloader-resolver';
import { initializeProxy } from './proxy';
import {
  withTempDirectoryIn,
  normalizeVersion,
  getHostArch,
  getNodeArch,
  ensureIsTruthyString,
  isOfficialLinuxIA32Download,
} from './utils';

export { getHostArch } from './utils';
export { initializeProxy } from './proxy';
export * from './types';

const d = debug('@electron/get:index');

if (process.env.ELECTRON_GET_USE_PROXY) {
  initializeProxy();
}

/**
 * Downloads a specific version of Electron and returns an absolute path to a
 * ZIP file.
 *
 * @param version - The version of Electron you want to download
 */
export function download(
  version: string,
  options?: ElectronDownloadRequestOptions,
): Promise<string> {
  return downloadArtifact({
    ...options,
    version,
    platform: process.platform,
    arch: process.arch,
    artifactName: 'electron',
  });
}

/**
 * Downloads an artifact from an Electron release and returns an absolute path
 * to the downloaded file.
 *
 * @param artifactDetails - The information required to download the artifact
 */
export async function downloadArtifact(
  _artifactDetails: ElectronPlatformArtifactDetailsWithDefaults,
): Promise<string> {
  const artifactDetails: ElectronArtifactDetails = _artifactDetails.isGeneric
    ? {
        ..._artifactDetails,
      }
    : {
        platform: process.platform,
        arch: process.arch,
        ..._artifactDetails,
      };
  ensureIsTruthyString(artifactDetails, 'version');
  artifactDetails.version = normalizeVersion(artifactDetails.version);
  if (artifactDetails.hasOwnProperty('arch')) {
    (artifactDetails as any).arch = getNodeArch((artifactDetails as any).arch);
  }

  const fileName = getArtifactFileName(artifactDetails);
  const url = getArtifactRemoteURL(artifactDetails);
  const cache = new Cache(artifactDetails.cacheRoot);

  // Do not check if the file exists in the cache when force === true
  if (!artifactDetails.force) {
    d(`Checking the cache (${artifactDetails.cacheRoot}) for ${fileName} (${url})`);
    const cachedPath = await cache.getPathForFileInCache(url, fileName);

    if (cachedPath === null) {
      d('Cache miss');
    } else {
      d('Cache hit');
      return cachedPath;
    }
  }

  if (
    !artifactDetails.isGeneric &&
    isOfficialLinuxIA32Download(
      artifactDetails.platform,
      artifactDetails.arch,
      artifactDetails.version,
      artifactDetails.mirrorOptions,
    )
  ) {
    console.warn('Official Linux/ia32 support is deprecated.');
    console.warn('For more info: https://electronjs.org/blog/linux-32bit-support');
  }

  return await withTempDirectoryIn(artifactDetails.tempDirectory, async tempFolder => {
    const tempDownloadPath = path.resolve(tempFolder, getArtifactFileName(artifactDetails));

    const downloader = artifactDetails.downloader || (await getDownloaderForSystem());
    d(
      `Downloading ${url} to ${tempDownloadPath} with options: ${JSON.stringify(
        artifactDetails.downloadOptions,
      )}`,
    );
    await downloader.download(url, tempDownloadPath, artifactDetails.downloadOptions);

    // Don't try to verify the hash of the hash file itself
    if (
      !artifactDetails.artifactName.startsWith('SHASUMS256') &&
      !artifactDetails.unsafelyDisableChecksums
    ) {
      const shasumPath = await downloadArtifact({
        isGeneric: true,
        version: artifactDetails.version,
        artifactName: 'SHASUMS256.txt',
        force: artifactDetails.force,
        downloadOptions: artifactDetails.downloadOptions,
        cacheRoot: artifactDetails.cacheRoot,
        downloader: artifactDetails.downloader,
        mirrorOptions: artifactDetails.mirrorOptions,
      });

      await sumchecker('sha256', shasumPath, path.dirname(tempDownloadPath), [
        path.basename(tempDownloadPath),
      ]);
    }

    return await cache.putFileInCache(url, tempDownloadPath, fileName);
  });
}
