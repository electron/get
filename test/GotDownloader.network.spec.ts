import * as fs from 'fs-extra';
import * as path from 'path';

import { GotDownloader } from '../src/GotDownloader';
import { withTempDirectory } from '../src/utils';
import { EventEmitter } from 'events';

describe('GotDownloader', () => {
  describe('download()', () => {
    it('should download a remote file to the given file path', async () => {
      const downloader = new GotDownloader();
      await withTempDirectory(async dir => {
        const testFile = path.resolve(dir, 'test.txt');
        expect(await fs.pathExists(testFile)).toEqual(false);
        await downloader.download(
          'https://github.com/electron/electron/releases/download/v2.0.18/SHASUMS256.txt',
          testFile,
        );
        expect(await fs.pathExists(testFile)).toEqual(true);
        expect(await fs.readFile(testFile, 'utf8')).toMatchSnapshot();
      });
    });

    it('should throw an error if the file does not exist', async function() {
      if (process.platform === 'win32') {
        return;
      }
      const downloader = new GotDownloader();
      await withTempDirectory(async dir => {
        const testFile = path.resolve(dir, 'test.txt');
        await expect(
          downloader.download(
            'https://github.com/electron/electron/releases/download/v2.0.18/bad.file',
            testFile,
          ),
        ).rejects.toMatchInlineSnapshot(`[HTTPError: Response code 404 (Not Found)]`);
      });
    });

    it('should throw an error if the file write stream fails', async () => {
      const downloader = new GotDownloader();
      const spy = jest.spyOn(fs, 'createWriteStream');
      spy.mockImplementationOnce(() => {
        const emitter = new EventEmitter();
        setTimeout(() => emitter.emit('error', 'bad write error thing'), 10);
        return emitter as any;
      });
      await withTempDirectory(async dir => {
        const testFile = path.resolve(dir, 'test.txt');
        await expect(
          downloader.download(
            'https://github.com/electron/electron/releases/download/v2.0.18/bad.file',
            testFile,
          ),
        ).rejects.toMatchInlineSnapshot(`"bad write error thing"`);
      });
    });

    it('should download to a deep uncreated path', async () => {
      const downloader = new GotDownloader();
      await withTempDirectory(async dir => {
        const testFile = path.resolve(dir, 'f', 'b', 'test.txt');
        expect(await fs.pathExists(testFile)).toEqual(false);
        await expect(
          downloader.download(
            'https://github.com/electron/electron/releases/download/v2.0.1/SHASUMS256.txt',
            testFile,
          ),
        ).resolves.toBeUndefined();
        expect(await fs.pathExists(testFile)).toEqual(true);
        expect(await fs.readFile(testFile, 'utf8')).toMatchSnapshot();
      });
    });
  });
});
