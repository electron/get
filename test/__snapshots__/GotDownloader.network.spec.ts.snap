// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GotDownloader download() should download a remote file to the given file path 1`] = `
"d186456c8e44155298521aafc5e270875c6cde41bbd3c19ce38ae53f1d98f284 *chromedriver-v2.0.18-darwin-x64.zip
dfb244c4347481a526567418f78a0fb6b9ccff6205a55f51225cfe4d1796b0d9 *chromedriver-v2.0.18-linux-arm.zip
516981ca6f7465dba55228df9ba30b7e0566a3eec694e5a2cc2fbe58a41ff006 *chromedriver-v2.0.18-linux-arm64.zip
dfb244c4347481a526567418f78a0fb6b9ccff6205a55f51225cfe4d1796b0d9 *chromedriver-v2.0.18-linux-armv7l.zip
6750ebb6f48e2ba104d33f8f407818d8316bcc438590bb04a2dfb03cd069f2f4 *chromedriver-v2.0.18-linux-ia32.zip
37140f6ec7d333dcd559c85815d547ef7a0046272f37fce0f78c308032779edc *chromedriver-v2.0.18-linux-x64.zip
081baba2d4c47f940ed19b101a53d9a576d7519cc9db8d15d1180190ce86b65c *chromedriver-v2.0.18-mas-x64.zip
794370f5ee270b83682a91c658092a3b3a71dbebfded9f48159dc162d32ad64b *chromedriver-v2.0.18-win32-ia32.zip
9c5ec68bb6e92ef6c61ffaf139d651e7760d79dc457b91fefa4b817f13fd2a9b *chromedriver-v2.0.18-win32-x64.zip
db086eeb6619cbf61e673f3bcd69a24df04e4b98f4cf458cf76e1f02eb2b7ee8 *electron-api.json
dbd50c1fd4bad524768243a3817bffc15a400779838099fb229d4861d3dba02b *electron-v2.0.18-darwin-x64-dsym.zip
3a829d3e0adc6994bdb27d818a9e9a8e404e2e3d5d4811db2f1a0c52f6a70742 *electron-v2.0.18-darwin-x64-symbols.zip
2055e52fc3a336c9c37dc0a03ca8559d62a4288bdf4160aed875e9dc57a0a2da *electron-v2.0.18-darwin-x64.zip
9d47382091dd917ade29d36af3e7a7ce9abcf1f76457c228ab8b4c59c7a02f2d *electron-v2.0.18-linux-arm-symbols.zip
75c37301e958368263cee27539bc80ab9bcf810a83f3e3dd6367f6646866b9fb *electron-v2.0.18-linux-arm.zip
7b3cca296b6f77b5c53488eca17ccda072ec09259431373067ffc026db11f4c4 *electron-v2.0.18-linux-arm64-symbols.zip
dfcfb60fc57dcec41a25cebbea37c0486a0ffd690ac3ccbb80a97161851df81f *electron-v2.0.18-linux-arm64.zip
9d47382091dd917ade29d36af3e7a7ce9abcf1f76457c228ab8b4c59c7a02f2d *electron-v2.0.18-linux-armv7l-symbols.zip
75c37301e958368263cee27539bc80ab9bcf810a83f3e3dd6367f6646866b9fb *electron-v2.0.18-linux-armv7l.zip
16f738a1a85eb0f3f916e52130300e37e88177e9da57b5bf32766f2ff1114773 *electron-v2.0.18-linux-ia32-symbols.zip
05bbd3e73776143ac357b0c25a631bf2e503a922b9506b23d636bd0b06a3f2f6 *electron-v2.0.18-linux-ia32.zip
1441f7cea2d3d1964f3ead1c7fd7020e83dd9d06bd716ef3a527011ecf0f39ba *electron-v2.0.18-linux-x64-symbols.zip
f196e06b6ecfa33bffb02b3d6c4a64bd5a076014e2f21c4a67356474ee014000 *electron-v2.0.18-linux-x64.zip
c06dd8931e0583c6989369924edc4c115050086a1fb41f4741aff186cf979955 *electron-v2.0.18-mas-x64-dsym.zip
b2f177134f11481c7ee4fd7fdcf027b699199d3b4c20b04bc3e31ac37937c2a9 *electron-v2.0.18-mas-x64-symbols.zip
25c20dbfcdab16e1ec3bd180bf845470ae0f4b784b4ef2e6dbba77fc5735258c *electron-v2.0.18-mas-x64.zip
03b82b19205ef6341b7f5fd6c7e90adb70c3e6207b83add00018b598602b3a2b *electron-v2.0.18-win32-ia32-pdb.zip
81f424eb0f318476a4fc878329f38c98666677b7cc507d27c54a11b61cf8580a *electron-v2.0.18-win32-ia32-symbols.zip
37fdb486fa3f0aea2678c4159f797176b02e15c50e81fc059e542fbd1f3ff0d7 *electron-v2.0.18-win32-ia32.zip
af2d527148be51f11503112b63546419b8cd15f9e59fd24bc3e758fb15b7cc31 *electron-v2.0.18-win32-x64-pdb.zip
09c6ed3d4a52c176dcdd8ebdaef25563575809fddec1b8243675cde2b731c39a *electron-v2.0.18-win32-x64-symbols.zip
904884aca1dec4c036243dfd3cffb04df62d1cdd290323551476729f41db50fc *electron-v2.0.18-win32-x64.zip
80c89e7028a4db7fd028050bac19bd90bdef336b7dc56eeb9d105ca83450689f *electron.d.ts
0bcff2dfa28da5dc5924e990daf9e22ac243bc58b4fe7a6225beba6a6da276f3 *ffmpeg-v2.0.18-darwin-x64.zip
5cce7303e185602857292c216d00d5726e63c50f36c1527238b85b9e80713d2f *ffmpeg-v2.0.18-linux-arm.zip
e9acaac5dce3f6eae23abb89121f8bafaed7595d8678700c0975643929b8dd4b *ffmpeg-v2.0.18-linux-arm64.zip
5cce7303e185602857292c216d00d5726e63c50f36c1527238b85b9e80713d2f *ffmpeg-v2.0.18-linux-armv7l.zip
32aa39326628207ddf5f0889b73f130e8e7c6c66a308e9a45156a7dc09045e60 *ffmpeg-v2.0.18-linux-ia32.zip
550c628a55670ceff174ff39826272c9501560ec7fa15eb4e0779efc26a00727 *ffmpeg-v2.0.18-linux-x64.zip
a577d484f257aa0e66f65a1f56b5e95cd3e5a7a97cb1a21a08a82fe8bb75dbc9 *ffmpeg-v2.0.18-mas-x64.zip
af0236e9443833c91816901737cd752b5ac03848b986d805c89ffdc1fc591585 *ffmpeg-v2.0.18-win32-ia32.zip
047b2db1c178cf76b161eb320b58e038bc5435927b0abb4ce425af999947ac7a *ffmpeg-v2.0.18-win32-x64.zip
d09a05aed8fd4ac7a91c9be250f66cb8bf8f1c22243f93f1bbbcc09b7b4edf50 *mksnapshot-v2.0.18-darwin-x64.zip
d2a8120b4da737a507d6b4a9c2cc1a53c32c89b7ea606cd68effc76b8356830a *mksnapshot-v2.0.18-linux-arm-x64.zip
788431928a32098f1a83d2b89c7863fbc33bb0ee02d34f4eaf08098a33600864 *mksnapshot-v2.0.18-linux-arm.zip
2fbd5f590d038dcff7980a0954cdda4bb4f580bc318f6af7dd58b91471f4424a *mksnapshot-v2.0.18-linux-arm64-x64.zip
18a20f3ef09e701266474728a7d2a17fb2edff0170eaaf757097a4d7d47b0715 *mksnapshot-v2.0.18-linux-arm64.zip
d2a8120b4da737a507d6b4a9c2cc1a53c32c89b7ea606cd68effc76b8356830a *mksnapshot-v2.0.18-linux-armv7l-x64.zip
788431928a32098f1a83d2b89c7863fbc33bb0ee02d34f4eaf08098a33600864 *mksnapshot-v2.0.18-linux-armv7l.zip
fed7fddc94c27d3232c63e76ea7b1248618712708fa9936cf492b4147227d739 *mksnapshot-v2.0.18-linux-ia32.zip
f83f0433d0e2f4d62db59892d4105c5693242da867cb76f3ec42ce020868bf98 *mksnapshot-v2.0.18-linux-x64.zip
68255b3aecd3f743a07338b6fc964d6e5fcd9ad37cdbd696b5ebcf7d57f27fd2 *mksnapshot-v2.0.18-mas-x64.zip
1dae0fc78ea703032833c0a2d535155ffbfd0b24ec5b894aa0e8aff2bcd9a1a8 *mksnapshot-v2.0.18-win32-ia32.zip
7cb445597f95c202a480b737b0df1d39eab044e37e090484494616017eb36a22 *mksnapshot-v2.0.18-win32-x64.zip
"
`;

exports[`GotDownloader download() should download to a deep uncreated path 1`] = `
"06fbdc5cdebf63b6fc579092d572878304125286b0db3ed0d6b1d506dc75e286 *chromedriver-v2.0.1-darwin-x64.zip
7347204e847fd7403dfac7de0a92a345932cc99a60b3279ca9e6d307483a190a *chromedriver-v2.0.1-linux-arm.zip
0a5807a4f26bc42de70d008e6ec775fed54a66f55d517b68071b801c3b306b4c *chromedriver-v2.0.1-linux-arm64.zip
7347204e847fd7403dfac7de0a92a345932cc99a60b3279ca9e6d307483a190a *chromedriver-v2.0.1-linux-armv7l.zip
c9060d1c384bcd3dcec43b56f43ef1f6668c789833d93175a85d9d4d79c176fa *chromedriver-v2.0.1-linux-ia32.zip
522d870210e453a1856ad2eb5db185749a8738e0ad63f6bbe2f0d03a4974fe6a *chromedriver-v2.0.1-linux-x64.zip
813db89ca0a8c449d1cc3571bbeb7dbef03e277620a5841c75fb816e1d300f6e *chromedriver-v2.0.1-mas-x64.zip
8b561b17b0917a50fa15a0c67271d93a899f5d60c9229bec59ddb94492c56f8c *chromedriver-v2.0.1-win32-ia32.zip
7c09430f50e7e578cf76cd355b47e74d031215e97d3d7b73135b1fd0601d52fa *chromedriver-v2.0.1-win32-x64.zip
1ffc7c7a51aae73783e6c7f8112fb1746033de47957ac672831a98f267cee727 *electron-api.json
0519c4d413be3dd5a482d4a02aef6025edcbe1fc55ce9bd26e4d0d688be10db1 *electron-v2.0.1-darwin-x64-dsym.zip
d6a87486f36dac598487b73ede4c4d8b257f87e4b9d69e50d1cc36b5a2f07eb2 *electron-v2.0.1-darwin-x64-symbols.zip
93037deeb1f0c4fe90913e8c4360c323badfc223d9c00d8a4e28dc9a474c60d2 *electron-v2.0.1-darwin-x64.zip
fc06151c32d221ec65a404a56e0a6caac9790fdf63d5422f3fe6c268877ff106 *electron-v2.0.1-linux-arm-symbols.zip
c31d48bc4a8a845c5edc300aad2c4114f174b5b828b8d5093850b62f5b86d0c1 *electron-v2.0.1-linux-arm.zip
e74e3465c925339d19ce7448fce0040c06efe3fba386583a77a959d6f48f3e5a *electron-v2.0.1-linux-arm64-symbols.zip
a61fb0a1f4cdb1126a3f43d90fbfb40c80c263f0091f548611fd91f0a22ab19b *electron-v2.0.1-linux-arm64.zip
fc06151c32d221ec65a404a56e0a6caac9790fdf63d5422f3fe6c268877ff106 *electron-v2.0.1-linux-armv7l-symbols.zip
c31d48bc4a8a845c5edc300aad2c4114f174b5b828b8d5093850b62f5b86d0c1 *electron-v2.0.1-linux-armv7l.zip
db2637d6cedd278e7f890bb6fffdeba630eab59726edbc072a0af64ff5df5019 *electron-v2.0.1-linux-ia32-symbols.zip
c6bfaaee82cbddf490288bb3dc3ac1e26be8c363899508027159411d818fa681 *electron-v2.0.1-linux-ia32.zip
76c1273ed5e2a2e4b0eb0de37c4e22a75106df5b683bc1581f554d632bade9ba *electron-v2.0.1-linux-x64-symbols.zip
c95e44ff13eb86bde49e909d64972794e23e124d00060d0e3516174448146049 *electron-v2.0.1-linux-x64.zip
b06e6c864c2a81e848d3f92222b56a451c4c1ce5759b792df8ba944cd769c6a9 *electron-v2.0.1-mas-x64-dsym.zip
815122b13828b07bffe7eedf0afc6f8aca71d03bc712e61b61689d56de1407e5 *electron-v2.0.1-mas-x64-symbols.zip
f31012bed957b688e3b74ed533bafe24622fab6435e1b4c17a90943fc4acf940 *electron-v2.0.1-mas-x64.zip
6856affa65de4ee43b6f1daec1d525ab8743807a6ac6224084597658f730281a *electron-v2.0.1-win32-ia32-pdb.zip
d5ac8f04f71b80ab9c0ac84bbed04622e5393aa44d85e29d8af83e224619ef53 *electron-v2.0.1-win32-ia32-symbols.zip
dac3116900550f72b92192ba3191b6b62968307d635167fd16f4ca0cb6df167e *electron-v2.0.1-win32-ia32.zip
b8501a49f573977a398d3133f46bebeff5b4e3727c1c708465c775023f268849 *electron-v2.0.1-win32-x64-pdb.zip
d3e4136a4db1c58c47fdb8cc3e5ea731cd48f60c372b43a216ead9aaf2c11324 *electron-v2.0.1-win32-x64-symbols.zip
d754696b75e75b54b18d08bf70ba4bd5b04c3330a44c4ddc371b8e39ad51e180 *electron-v2.0.1-win32-x64.zip
d7dff9f5df46dd0459a2203b2ce043f37c70369eec122b81631d5f49316a8e2b *electron.d.ts
8778c206570b68455a79844b569ec3a6017387d479bea9329672224f9c78b410 *ffmpeg-v2.0.1-darwin-x64.zip
aebfa31700b637594aaebd18cfb810953998195bfc1b0b6f3e55e7712699afd6 *ffmpeg-v2.0.1-linux-arm.zip
c41546e75b4e23526d4eb040ebe2d24620dc9feacbc2f1999da327e7acd0a0b2 *ffmpeg-v2.0.1-linux-arm64.zip
aebfa31700b637594aaebd18cfb810953998195bfc1b0b6f3e55e7712699afd6 *ffmpeg-v2.0.1-linux-armv7l.zip
9d4b233d7fbf7d49ba2b85eb5f7e86ff2487616d27388ac84673f24e20aae275 *ffmpeg-v2.0.1-linux-ia32.zip
711c36c0a845b7e5901580f5b6ddbef21082faaa03f103bd423787dcab74347f *ffmpeg-v2.0.1-linux-x64.zip
05536cb36883a9eb10d112e18b201b09d097274d88c41d44fa54a5cbd5f3c638 *ffmpeg-v2.0.1-mas-x64.zip
9f49ad8e266677bd7c8a077967900d3b01fd88e4a438f13eb04e8dc24769ac8a *ffmpeg-v2.0.1-win32-ia32.zip
4c17b21f05e85a30e7d4e7e7d25471caeacde33b6300cbb08141d02fd827d7b5 *ffmpeg-v2.0.1-win32-x64.zip
34fb1abb38aa286a95ae74ce83dc48c71ddd83d2fdcc2201bdc6890abfb175ee *mksnapshot-v2.0.1-darwin-x64.zip
ad510fb748dcdf8b2b189f220a8720bd358580c27e8a80cec6377bd56f63faac *mksnapshot-v2.0.1-linux-arm-x64.zip
6f4e3bb996ab9355914bc96aa3fc69afd2c5145525006898fbae633b94206ac1 *mksnapshot-v2.0.1-linux-arm.zip
dc96949c678732e25212aea2e784f31a75d4f03f1e1fdede6122a27412b7d3e8 *mksnapshot-v2.0.1-linux-arm64-x64.zip
dae0d49148910b07a563e01408611f767e50c1d979dbc77b67426be429c6e9fd *mksnapshot-v2.0.1-linux-arm64.zip
ad510fb748dcdf8b2b189f220a8720bd358580c27e8a80cec6377bd56f63faac *mksnapshot-v2.0.1-linux-armv7l-x64.zip
6f4e3bb996ab9355914bc96aa3fc69afd2c5145525006898fbae633b94206ac1 *mksnapshot-v2.0.1-linux-armv7l.zip
ef075a31c85c41fb1ab1628a8f811f90354769652b17f0e01d40fa6e0678b558 *mksnapshot-v2.0.1-linux-ia32.zip
d555686b04db329b67d2cf133ff83e4d566e0f2a99388bdf3eb2d0b73b5d34bd *mksnapshot-v2.0.1-linux-x64.zip
5e6799a4516044d7d77988b9668e27d5c00516ff87dd382b7a9bdf9501e7b7ee *mksnapshot-v2.0.1-mas-x64.zip
4de5336d82e67b0db803b9273ff617637248a67b0980a4d6263b4945ccf56557 *mksnapshot-v2.0.1-win32-ia32.zip
964de300a37438cd5259a7de494fdee07d3faec0b694b438d27519b6677d9961 *mksnapshot-v2.0.1-win32-x64.zip
"
`;
